File: server/src/routes/skills.ts
import { Router } from 'express';
import { query } from '../database/index.js';
import { authenticate, AuthRequest } from '../middleware/auth.js';
import { ValidationError } from '../middleware/errorHandler.js';

const router = Router();

// Get skills for a candidate
router.get('/candidate/:candidateId', authenticate, async (req: AuthRequest, res, next) => {
  try {
    const { candidateId } = req.params;
    const result = await query(
      'SELECT * FROM skills WHERE candidate_id = $1 ORDER BY created_at DESC',
      [candidateId]
    );

    res.json({
      success: true,
      data: result.rows
    });
  } catch (error) {
    next(error);
  }
});

// Get single skill
router.get('/:id', authenticate, async (req: AuthRequest, res, next) => {
  try {
    const { id } = req.params;
    const result = await query('SELECT * FROM skills WHERE id = $1', [id]);

    if (result.rows.length === 0) {
      throw new ValidationError('Skill not found');
    }

    res.json({
      success: true,
      data: result.rows[0]
    });
  } catch (error) {
    next(error);
  }
});

// Create skill
router.post('/', authenticate, async (req: AuthRequest, res, next) => {
  try {
    const {
      candidate_id,
      name,
      category,
      resume_claim_level
    } = req.body;

    if (!candidate_id || !name || !category) {
      throw new ValidationError('Missing required fields');
    }

    const result = await query(
      `INSERT INTO skills (candidate_id, name, category, resume_claim_level)
      VALUES ($1, $2, $3, $4)
      RETURNING *`,
      [candidate_id, name, category, resume_claim_level]
    );

    res.status(201).json({
      success: true,
      data: result.rows[0]
    });
  } catch (error) {
    next(error);
  }
});

// Update skill
router.put('/:id', authenticate, async (req: AuthRequest, res, next) => {
  try {
    const { id } = req.params;
    const {
      evidence_count,
      verification_state,
      evidence_strength,
      missing_proof_reason,
      recommended_validation,
      is_proven
    } = req.body;

    const result = await query(
      `UPDATE skills SET
        evidence_count = COALESCE($1, evidence_count),
        verification_state = COALESCE($2, verification_state),
        evidence_strength = COALESCE($3, evidence_strength),
        missing_proof_reason = COALESCE($4, missing_proof_reason),
        recommended_validation = COALESCE($5, recommended_validation),
        is_proven = COALESCE($6, is_proven),
        updated_at = CURRENT_TIMESTAMP
      WHERE id = $7
      RETURNING *`,
      [evidence_count, verification_state, evidence_strength, missing_proof_reason, recommended_validation, is_proven, id]
    );

    if (result.rows.length === 0) {
      throw new ValidationError('Skill not found');
    }

    res.json({
      success: true,
      data: result.rows[0]
    });
  } catch (error) {
    next(error);
  }
});

export { router as skillsRouter };

File: backend/app/routes/skills.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from app.database import get_db
from app.models.skill import Skill
from app.models.user import User
from app.schemas import SkillCreate, SkillUpdate, ApiResponse
from app.auth import get_current_active_user

router = APIRouter()


@router.get("/candidate/{candidate_id}", response_model=ApiResponse)
async def get_skills_by_candidate(
    candidate_id: str,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    skills = db.query(Skill).filter(Skill.candidate_id == candidate_id).order_by(Skill.created_at.desc()).all()
    
    return ApiResponse(
        success=True,
        data={
            "skills": [{
                "id": str(s.id),
                "candidate_id": str(s.candidate_id),
                "name": s.name,
                "category": s.category,
                "resume_claim_level": s.resume_claim_level,
                "evidence_count": s.evidence_count,
                "verification_state": s.verification_state,
                "evidence_strength": s.evidence_strength,
                "missing_proof_reason": s.missing_proof_reason,
                "recommended_validation": s.recommended_validation,
                "is_proven": s.is_proven,
                "created_at": s.created_at.isoformat(),
                "updated_at": s.updated_at.isoformat()
            } for s in skills]
        }
    )


@router.get("/{skill_id}", response_model=ApiResponse)
async def get_skill(
    skill_id: str,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    skill = db.query(Skill).filter(Skill.id == skill_id).first()
    if not skill:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Skill not found"
        )
    
    return ApiResponse(
        success=True,
        data={
            "skill": {
                "id": str(skill.id),
                "candidate_id": str(skill.candidate_id),
                "name": skill.name,
                "category": skill.category,
                "resume_claim_level": skill.resume_claim_level,
                "evidence_count": skill.evidence_count,
                "verification_state": skill.verification_state,
                "evidence_strength": skill.evidence_strength,
                "missing_proof_reason": skill.missing_proof_reason,
                "recommended_validation": skill.recommended_validation,
                "is_proven": skill.is_proven,
                "created_at": skill.created_at.isoformat(),
                "updated_at": skill.updated_at.isoformat()
            }
        }
    )


@router.post("/", response_model=ApiResponse)
async def create_skill(
    skill_data: SkillCreate,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    if not skill_data.candidate_id or not skill_data.name or not skill_data.category:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Missing required fields"
        )
    
    new_skill = Skill(**skill_data.model_dump())
    db.add(new_skill)
    db.commit()
    db.refresh(new_skill)
    
    return ApiResponse(
        success=True,
        data={
            "skill": {
                "id": str(new_skill.id),
                "candidate_id": str(new_skill.candidate_id),
                "name": new_skill.name,
                "category": new_skill.category,
                "resume_claim_level": new_skill.resume_claim_level,
                "evidence_count": new_skill.evidence_count,
                "verification_state": new_skill.verification_state,
                "evidence_strength": new_skill.evidence_strength,
                "missing_proof_reason": new_skill.missing_proof_reason,
                "recommended_validation": new_skill.recommended_validation,
                "is_proven": new_skill.is_proven,
                "created_at": new_skill.created_at.isoformat(),
                "updated_at": new_skill.updated_at.isoformat()
            }
        }
    )


@router.put("/{skill_id}", response_model=ApiResponse)
async def update_skill(
    skill_id: str,
    skill_data: SkillUpdate,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    skill = db.query(Skill).filter(Skill.id == skill_id).first()
    if not skill:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Skill not found"
        )
    
    # Update fields
    update_data = skill_data.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(skill, field, value)
    
    db.commit()
    db.refresh(skill)
    
    return ApiResponse(
        success=True,
        data={
            "skill": {
                "id": str(skill.id),
                "candidate_id": str(skill.candidate_id),
                "name": skill.name,
                "category": skill.category,
                "resume_claim_level": skill.resume_claim_level,
                "evidence_count": skill.evidence_count,
                "verification_state": skill.verification_state,
                "evidence_strength": skill.evidence_strength,
                "missing_proof_reason": skill.missing_proof_reason,
                "recommended_validation": skill.recommended_validation,
                "is_proven": skill.is_proven,
                "created_at": skill.created_at.isoformat(),
                "updated_at": skill.updated_at.isoformat()
            }
        }
    )
